### PR TITLE
moving code out of the crowdsale contract

### PR DESCRIPTION
**Pull request by moodysalem**
have a look and let me know if you have any questions-the big goal is to remove code from the crowdsale contract that can be better encapsulated in testable pieces. e.g. the `AddressWhitelist` and  `KnowsConstants` contract

### CODE REVIEW

File: contracts/AddressWhitelist.sol [hunk 1]
```diff
@@ -0,0 +1,43 @@
+pragma solidity ^0.4.18;
+
+import 'zeppelin-solidity/contracts/ownership/Ownable.sol';
+
+// A simple contract that stores a whitelist of addresses, which the owner may update
+contract AddressWhitelist is Ownable {
```
Comment (dnzdlklc) [inline contracts/AddressWhitelist.sol]: Really like this separation of whitelisting. 👍

File: contracts/BntyExchangeRateCalculator.sol [hunk 2]
```diff
@@ -0,0 +1,36 @@
+pragma solidity ^0.4.18;
+
+import 'zeppelin-solidity/contracts/math/SafeMath.sol';
+
+// This contract does the math to figure out the BNTY paid per WEI, based on the fixed USD ether price
+contract BntyExchangeRateCalculator {
+    using SafeMath for uint;
+
+    // a microdollar is one millionth of a dollar, or one ten-thousandth of a cent
+    uint public constant MICRO_DOLLARS_PER_BNTY_PRESALE = 16500;
```
Comment (dnzdlklc) [inline contracts/BntyExchangeRateCalculator.sol]: I haven't seen this been used anywhere. Maybe its on `to-do list` to use it on `PreSaleDistributor` contract?

File: contracts/KnowsConstants.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    uint public constant HARD_CAP_AMOUNT_USD = 1500000;                           // in USD the maximum total collected amount
+    uint public constant MAXIMUM_CONTRIBUTION_AMOUNT_USD_DURING_WHITELIST = 1500; // in USD the maximum contribution amount during the whitelist period
+    uint public constant MAXIMUM_CONTRIBUTION_AMOUNT_USD_POST_WHITELIST = 10000;  // in USD the maximum contribution amount after the whitelist period ends
+    uint public constant MAX_GAS_PRICE = 30 * (10 ** 9);                          // Max gas price of 30 gwei
+
+    // Time constants
+    uint public constant SALE_START_DATE = 1513346400;                    // in unix timestamp Dec 15th @ 15:00 CET
+    uint public constant WHITELIST_END_DATE = SALE_START_DATE + 24 hours; // End whitelist 24 hours after sale start date/time
+    uint public constant SALE_END_DATE = SALE_START_DATE + 4 weeks;       // end sale in four weeks
+    uint public constant UNFREEZE_DATE = SALE_START_DATE + 76 weeks;      // Bounty0x Reserve locked for 18 months
```
Comment (dnzdlklc) [inline contracts/KnowsConstants.sol]: When implementing this have a look at https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/token/TokenTimelock.sol

However we'd need additional feature to unfreeze and pull some tokens from it i.e. to onboard a team member and transfer `x` number of tokens to the new members account as part of compensation.

File: migrations/2_deploy_contracts.js [hunk 4]
```diff
@@ -2,12 +2,12 @@ const Bounty0xToken = artifacts.require('Bounty0xToken');
 const Bounty0xCrowdsale = artifacts.require('Bounty0xCrowdsale');
 const MiniMeTokenFactory = artifacts.require('MiniMeTokenFactory');
 
-const founder1 = '0x60d7df77bcc92a0e92c6d2b7b4d276ad0dd33e90';
-const founder2 = '0xd32ea3da0044fc5c9554a43bbbb3899c0124a9b5';
-const founder3 = '0xb2427291cb661a2ed72c1e66a9fe2faffbb67b2f';
-const bounty0Wallet = '0xc9afbf88b36a5c4a0a8a41918552e24a5c3a1958';
+const FOUNDER_1 = '0x60d7df77bcc92a0e92c6d2b7b4d276ad0dd33e90';
```
Comment (dnzdlklc) [inline migrations/2_deploy_contracts.js]: These are all fake addresses. I will provide real addresses closer to main sale.

File: contracts/Bounty0xCrowdsale.sol [hunk 5]
```diff
@@ -6,104 +6,92 @@ import 'zeppelin-solidity/contracts/math/SafeMath.sol';
...
 
 import './Bounty0xToken.sol';
+import './BntyExchangeRateCalculator.sol';
+import './KnowsConstants.sol';
+import './AddressWhitelist.sol';
 import './Bounty0xPresaleDistributor.sol';
 
-contract Bounty0xCrowdsale is Pausable, TokenController {
-    using SafeMath for uint256;
+contract Bounty0xCrowdsale is KnowsConstants, BntyExchangeRateCalculator, AddressWhitelist, Pausable, TokenController {
```
Comment (dnzdlklc) [inline contracts/Bounty0xCrowdsale.sol]: importing of `Ownable` contract and making this contract extend `Ownable`

File: contracts/Bounty0xCrowdsale.sol [hunk 6]
```diff
@@ -145,51 +137,34 @@ contract Bounty0xCrowdsale is Pausable, TokenController {
...
-        return true;
-    }
-
-    // create the vesting contracts for the founder tokens
-    function distributeFounderTokens() public onlyOwner returns (bool success) {
-        require(vestingContracts[founder1] == address(0));
-        require(vestingContracts[founder2] == address(0));
-        require(vestingContracts[founder3] == address(0));
+    // This function call creates all the the vesting contracts and gives them the appropriate amounts of BNTY
+    function distributeVestedTokens() public onlyOwner returns (bool success) {
```
Comment (dnzdlklc) [inline contracts/Bounty0xCrowdsale.sol]: I'll look into this myself to see if `VestedToken` contract has such capability but we need to be able to stop vesting to a certain individual in the event person `x` leaves Bounty0x.

File: contracts/KnowsConstants.sol [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    uint public constant MAINSALE_POOL = 90900000;                        // 90.9M BNTY for Main-Sale Pool
+    uint public constant FOUNDER1_STAKE = 60000000;                       // 60M BNTY
+    uint public constant FOUNDER2_STAKE = 45000000;                       // 45M BNTY
+    uint public constant FOUNDER3_STAKE = 45000000;                       // 45M BNTY
+    uint public constant BOUNTY0X_RESERVE = 225150000;                    // 225.15M BNTY Bounty0x Reserve Pool
+    uint public constant ADVISERS_POOL = 15000000;                        // 15M BNTY Advisers Pool
+
+    // Vesting conditions
+    uint public constant TEAM_VESTING_CLIFF = 0 weeks;                  // no cliff
+    uint public constant TEAM_VESTING_PERIOD = 52 weeks;                // 1 year vesting period for founders
```
Comment (dnzdlklc) [inline contracts/KnowsConstants.sol]: In light of our ICO structure this has changed to 18 months.

File: contracts/BntyExchangeRateCalculator.sol (hunk 2 above)
Comment (moodysalem) [reply contracts/BntyExchangeRateCalculator.sol]: The main contract inherits from it and uses it in contribute

File: contracts/KnowsConstants.sol (hunk 3 above)
Comment (moodysalem) [reply contracts/KnowsConstants.sol]: This contract does have the ability to be locked

File: contracts/Bounty0xCrowdsale.sol (hunk 5 above)
Comment (moodysalem) [reply contracts/Bounty0xCrowdsale.sol]: Pausable  is Ownable but I could make it explicit

File: contracts/KnowsConstants.sol (hunk 7 above)
Comment (moodysalem) [reply contracts/KnowsConstants.sol]: Ok will change

File: contracts/KnowsConstants.sol (hunk 3 above)
Comment (moodysalem) [reply contracts/KnowsConstants.sol]: Oops, I thought we were talking about vested token. Let me think about this. I think if we are ok with the vesting period we can write a contract that let’s you deploy a new vested token with some of the reserve balance

File: contracts/Bounty0xCrowdsale.sol (hunk 5 above)
Comment (dnzdlklc) [reply contracts/Bounty0xCrowdsale.sol]: Lets make it explicit.

File: contracts/KnowsConstants.sol (hunk 3 above)
Comment (dnzdlklc) [reply contracts/KnowsConstants.sol]: That will work but wouldn't be easier to extend `TokenTimelock` contract and add additional functionality to allow withdrawal from it by `Owner` only?

Comment (moodysalem) [reply contracts/KnowsConstants.sol]: You're right, I made an additional commit that starts that out. Thanks

**Pull request merged**